### PR TITLE
TTSKit: decode-path performance improvements for Qwen3 streaming

### PR DESCRIPTION
**Pull request by mjfrey**
Four small optimizations to the Qwen3 TTS decode hot path, found while profiling on-device streaming synthesis in a production app (A19 iPhone, 12Hz 0.6B models):

- **Stateful decoders no longer copy K/V updates into the host-side cache.** For stateful models the MLState buffers already received this step's key/value updates during prediction, but the decode path still materialized the update tensors into the host-side cache arrays on every step. Only the cache position and attention masks need to advance. Adds `KVCache.update()` for that, with a unit test covering the position/mask bookkeeping.
- **K/V tensor materialization is concurrent.** The key and value MLTensor outputs are independent, but the async update paths materialized them sequentially — two round trips per decode step. `async let` runs both `toMLMultiArray()` calls concurrently in `update(keyTensor:valueTensor:)`, `updateStateCache`, and `updateWithHiddenContext`.
- **`BinaryDistinctString.init(NSString)` copies UTF-16 code units directly.** Converting through `Array(String)` walks Swift extended grapheme clusters and re-expands them to UTF-16 — needlessly expensive while loading large tokenizer vocabularies, and it can lose the exact NSString representation the type exists to preserve. `getCharacters(_:range:)` copies the code units as-is.
- **Top-k sampling selects the top-k logits before softmax** instead of computing probabilities over the entire vocabulary each step. Equivalent after renormalization, and avoids materializing full-vocabulary probabilities on the host.

## Measurements

A/B on an Apple silicon Mac (CLI, v1.0.0 vs v1.0.0 + these changes only, same seed/text/models, W8A16 0.6B, temperature 0.7 / top-k 20, 3 runs each):

| Metric | Before | After | Δ |
|---|---|---|---|
| Tokenizer load (Qwen3 151k vocab) | 0.95 s | 0.64 s | **−32%** |
| MultiCodeDecoder sampling | 0.91 ms/step | 0.81 ms/step | **−11%** |
| Total process CPU (16.5 s generation) | ~9.9 s | ~9.6 s | −2.5% |
| Throughput / RTF / peak RSS | — | — | unchanged |

End-to-end throughput is unchanged on the Mac because generation there is ANE-bound; the host-side savings are aimed at phone-class devices, where the decode loop shares CPU and memory bandwidth with the host app. The original motivation was an Instruments capture on an A19 iPhone during streaming synthesis, which attributed ~16% of captured cycles to the sampling path that these changes shrink.

All changes are behavior-preserving. `swift build` and the TTSKit unit tests (86) pass; the changes have been validated in continuous day-to-day on-device streaming use.

### CODE REVIEW

**Comment by ZachNagengast** [comment]
Thanks for the submission @mjfrey! Were you able to quantify the optimizations in terms of speed up or memory usage?

**Comment by mjfrey** [comment]
> Thanks for the submission @mjfrey! Were you able to quantify the optimizations in terms of speed up or memory usage?

@ZachNagengast Yes — A/B on an Apple silicon Mac (CLI, v1.0.0 vs v1.0.0 + these changes only, same seed/text/models, W8A16 0.6B, temperature 0.7 / top-k 20, 3 runs each):

| Metric | Before | After | Δ |
|---|---|---|---|
| Tokenizer load (151k vocab) | 0.95 s | 0.64 s | **−32%** (consistent across runs) |
| MultiCodeDecoder sampling | 0.91 ms/step | 0.81 ms/step | **−11%** |
| Total process CPU (16.5 s generation) | ~9.9 s | ~9.6 s | −2.5% |
| Throughput / RTF / peak memory | 19.8 steps/s, 0.64, ~470 MB | same | unchanged |

Throughput and memory are unchanged on the Mac because generation there is ANE-bound — these are host-CPU savings, aimed at phone-class devices where the decode loop shares CPU and memory bandwidth with the host app. The original motivation was an Instruments capture on an A19 iPhone during streaming synthesis, which attributed ~16% of captured cycles to the sampling path.

File: Sources/TTSKit/Qwen3TTS/Qwen3CodeDecoder.swift [hunk 1]
```diff
@@ -118,7 +118,13 @@ public class Qwen3CodeDecoder: CodeDecoding, @unchecked Sendable {
         }
 
         let cacheUpdateStart = CFAbsoluteTimeGetCurrent()
-        await cache.update(keyTensor: keyTensor, valueTensor: valueTensor)
+        if let _ = state as? MLState, isStateful {
+            // Stateful MLState already received the materialized outputs above;
+            // only advance the host-side position and masks here.
+            cache.update()
+        } else {
```
[the remark this answers is not in the document]
Comment (mjfrey) [reply Sources/TTSKit/Qwen3TTS/Qwen3CodeDecoder.swift]: Good catch — added `cache.isStateful` to the fast-path condition in 802b274, so a mismatched external-cache KVCache stays on the materializing path.

File: Sources/TTSKit/Utilities/Sampling.swift [hunk 2]
```diff
@@ -223,6 +223,27 @@ public class GreedyTokenSampler: TokenSampling, @unchecked Sendable {
...
+            // avoids computing probabilities for the entire vocabulary.
+            let scaledLogits = headLogits.cast(to: Float.self) / temperature
+            let (topKLogits, topKIndices) = scaledLogits.topK(topK)
+            let logitsArray = await topKLogits.toFloatArray()
+            let idxArray = await topKIndices.toIntArray()
+            let maxLogit = logitsArray.max() ?? 0
+            let weights = logitsArray.map { exp($0 - maxLogit) }
+            let weightSum = weights.reduce(0, +)
+            let randomValue = Float.random(in: 0..<weightSum, using: &rng)
+            var cumulativeWeight: Float = 0
```
[the remark this answers is not in the document]
Comment (mjfrey) [reply Sources/TTSKit/Utilities/Sampling.swift:245]: Added a guard in 802b274: when the weight sum is zero or non-finite the sampler now returns the top-scoring index deterministically (topK values are sorted descending) instead of trapping in `Float.random`.

File: Tests/TTSKitTests/TTSKitUnitTests.swift [hunk 3]
```diff
@@ -204,6 +204,19 @@ final class TTSKitUnitTests: XCTestCase {
...
     }
 
+    func testStatefulKVCacheUpdateWithoutTensorsAdvancesPositionAndMasks() throws {
+        let cache = try KVCache(cacheDim: 4, maxSeqLength: 8, isStateful: true)
+
+        cache.update()
+
+        XCTAssertEqual(cache.cacheLength, 1)
+        let updateMask = cache.kvCacheUpdateMask.dataPointer.bindMemory(to: FloatType.self, capacity: 8)
+        let paddingMask = cache.keyPaddingMask.dataPointer.bindMemory(to: FloatType.self, capacity: 8)
```
[the remark this answers is not in the document]
Comment (mjfrey) [reply Tests/TTSKitTests/TTSKitUnitTests.swift]: Done in 802b274 — the test now binds with the arrays' actual `count`.
